From fa336d92699c944b7f6128e41935b2281efbcc89 Mon Sep 17 00:00:00 2001
From: Charles Meng <charlesmeng@chromium.org>
Date: Thu, 22 Jan 2026 14:04:53 -0800
Subject: [PATCH] [Vertical Tabs] Always enlarge the discard indicator for vertical tabs

Update TabIcon so that it by default enlarges the discard indicator
radius unless its parent view notifies it that there is not enough
width to do so. This makes it so that the indicator radius is always
enlarged in the case of vertical tabs.

Comparison screenshots:
https://screenshot.googleplex.com/hJ8bcEqJ6VwGYyj.png
https://screenshot.googleplex.com/5PCEYzUgzjBRvHy.png

Bug: 475302379
Change-Id: Iaa899a268bfba42578dc2e962f8459645d3aa918
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/7510237
Reviewed-by: Eshwar Stalin <estalin@chromium.org>
Commit-Queue: Charles Meng <charlesmeng@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1573301}
---

--- a/chrome/browser/ui/views/tabs/tab.cc
+++ b/chrome/browser/ui/views/tabs/tab.cc
@@ -128,10 +128,6 @@ constexpr int kPinnedTabExtraWidthToRend
 constexpr int kTabAlertIndicatorCloseButtonPaddingAdjustmentTouchUI = 8;
 constexpr int kTabAlertIndicatorCloseButtonPaddingAdjustment = 4;
 
-// When the DiscardRingImprovements feature is enabled, increase the radius of
-// the discard ring by this amount if there is enough space.
-constexpr int kIncreasedDiscardIndicatorRadiusDp = 2;
-
 bool g_show_hover_card_on_mouse_hover = true;
 
 // Helper functions ------------------------------------------------------------
@@ -383,11 +379,8 @@ void Tab::Layout(PassKey) {
     } else {
       MaybeAdjustLeftForPinnedTab(&favicon_bounds, gfx::kFaviconSize);
     }
-    icon_->EnlargeDiscardIndicatorRadius(
-        width() - 2 * tab_style()->GetBottomCornerRadius() >=
-                gfx::kFaviconSize + 2 * kIncreasedDiscardIndicatorRadiusDp
-            ? kIncreasedDiscardIndicatorRadiusDp
-            : 0);
+    icon_->ResizeDiscardIndicatorRadiusForWidth(
+        width() - 2 * tab_style()->GetBottomCornerRadius());
 
     // Add space for insets outside the favicon bounds.
     favicon_bounds.Inset(-icon_->GetInsets());
--- a/chrome/browser/ui/views/tabs/tab_icon.cc
+++ b/chrome/browser/ui/views/tabs/tab_icon.cc
@@ -102,9 +102,8 @@ TabIcon::TabIcon()
   SetCanProcessEventsWithinSubtree(false);
 
   // Add padding to avoid clipping the attention indicator and the increased
-  // discard ring radius when kDiscardRingImprovements is enabled. Padding must
-  // be symmetric on each side so that elements will anchor to the center of the
-  // favicon.
+  // discard ring radius. Padding must be symmetric on each side so that
+  // elements will anchor to the center of the favicon.
   SetBorder(views::CreateEmptyBorder(kAttentionIndicatorRadius));
 
   const int preferred_width = gfx::kFaviconSize + GetInsets().width();
@@ -201,9 +200,11 @@ void TabIcon::StepLoadingAnimation(const
   }
 }
 
-void TabIcon::EnlargeDiscardIndicatorRadius(int radius) {
-  CHECK(radius <= GetInsets().left());
-  increased_discard_indicator_radius_ = radius;
+void TabIcon::ResizeDiscardIndicatorRadiusForWidth(int width) {
+  increased_discard_indicator_radius_ =
+      width >= gfx::kFaviconSize + 2 * kIncreasedDiscardIndicatorRadiusDp
+          ? kIncreasedDiscardIndicatorRadiusDp
+          : 0;
 }
 
 void TabIcon::OnDiscardRingTreatmentEnabledChanged() {
@@ -317,11 +318,10 @@ void TabIcon::PaintDiscardRingAndIcon(gf
   // Fades in the discard ring and smaller favicon
   MaybePaintFavicon(canvas, icon, icon_bounds);
 
-  // Increase the bounds of the discard ring beyond the icon bounds if
-  // kDiscardRingImprovements is enabled. This is safe because in the
-  // constructor, we have already added insets so that the larger discard ring
-  // can expand into them and won't be clipped, and the icon bounds will be
-  // inside those insets.
+  // Increase the bounds of the discard ring beyond the icon bounds if there is
+  // enough space in the tab. This is safe because in the constructor, we have
+  // already added insets so that the larger discard ring can expand into them
+  // and won't be clipped, and the icon bounds will be inside those insets.
   gfx::Rect discard_ring_bounds = icon_bounds;
   discard_ring_bounds.Outset(increased_discard_indicator_radius_);
 
--- a/chrome/browser/ui/views/tabs/tab_icon.h
+++ b/chrome/browser/ui/views/tabs/tab_icon.h
@@ -83,7 +83,11 @@ class TabIcon : public views::View, publ
   gfx::ImageSkia GetThemedIconForTesting() { return themed_favicon_; }
   bool GetActiveStateForTesting() { return is_active_tab_; }
 
-  void EnlargeDiscardIndicatorRadius(int radius);
+  // Called by the parent view when the width available to draw the tab icon
+  // changes, to determine whether to increase the discard indicator radius.
+  // Not used by vertical tabs since there always is enough width to draw the
+  // increased discard indicator radius.
+  void ResizeDiscardIndicatorRadiusForWidth(int width);
   void OnDiscardRingTreatmentEnabledChanged();
 
  protected:
@@ -211,7 +215,12 @@ class TabIcon : public views::View, publ
 
   bool is_monochrome_favicon_ = false;
 
-  int increased_discard_indicator_radius_ = 0;
+  // If there is enough space, increased the size of the discard ring beyond the
+  // default favicon size, by this amount.
+  static constexpr int kIncreasedDiscardIndicatorRadiusDp = 2;
+
+  // The amount to increase the size of the discard ring by.
+  int increased_discard_indicator_radius_ = kIncreasedDiscardIndicatorRadiusDp;
 
   PrefChangeRegistrar local_state_registrar_;
 };
